### PR TITLE
Make no. of services in load test configurable

### PR DESCRIPTION
**Pull request by shyamjvs**
Experimenting with https://github.com/kubernetes/kubernetes/issues/48107#issuecomment-398018717

/cc @wojtek-t 

```release-note
NONE
```

/kind bug
/sig scalability
/priority important-soon

### CODE REVIEW

File: test/e2e/scalability/load.go [hunk 1]
```diff
@@ -535,8 +535,11 @@ func GenerateConfigsForGroup(
 			MemRequest:     26214400, // 25MB
 			SecretNames:    secretNames,
 			ConfigMapNames: configMapNames,
-			// Define a label to group every 2 RCs into one service.
-			Labels: map[string]string{svcLabelKey: groupName + "-" + strconv.Itoa((i+1)/2)},
+		}
+
+		// Add a label to alternate RCs, to include only half of them in services.
+		if i%2 == 0 {
```
Comment (wojtek-t) [inline test/e2e/scalability/load.go]: I know that we want to experiment with that, but I don't want that to be default - this may lead to regressing e.g. how much traffic apiserver can handle.

I know that currently iptables cannot handled that many backends, but we at least know what apiserver can. I don't want to regress on that.

What we may do is to make that somehow flag-controllable (with default to all) and change that in one of the jobs?

Alternatively - can you a manual test with that (I can imagine that it doesn't even need to be 5k-node test for that purpose - you can run manual test on 2k-node test with 75pods/node or sth like that.

Comment (shyamjvs) [reply test/e2e/scalability/load.go]: That's a good point. I parametrized the test now wrt RC:Services ratio - so we should be able to easily experiment with different values.

> can you a manual test with that

I want to avoid doing it manually each time I tweak the value :) Also want the cluster logs recorded in gcs.

>  you can run manual test on 2k-node test with 75pods/node or sth like that

That's a good idea. But maybe we shouldn't change other params at the same time (like #watches, master size, #pods/node, etc) to avoid potentially different influences. Wdyt?

Comment (wojtek-t) [reply test/e2e/scalability/load.go]: > I want to avoid doing it manually each time I tweak the value :) Also want the cluster logs recorded in gcs.

>> you can run manual test on 2k-node test with 75pods/node or sth like that

> That's a good idea. But maybe we shouldn't change other params at the same time (like #watches, master size, #pods/node, etc) to avoid potentially different influences. Wdyt?

Yes - we shouldn't change to many params at the same time in general.
However, in this particular case, your goal is to only discover the limits of iptables, and for that particular purpose the smaller cluster we can achieve it with, the better.

I buy the argument why do you don't want to run it manually. But I also don't think that changing our main (release-blocking) tests for that purpose is the best idea.

What I started thinking about is that:
- we have this "manual-start" test-suite (which we are not utilizing at all)
- maybe what we can do is start utilizing it for exactly that purpose
- e.g. experiment would look like:
(a) send a PR to modify this manual test (to do what you want)
(b) manually trigger that job
(c) [optionally] revert that PR

What that would give us is:
- having logs in GCS
- ensuring that exactly one jobs is running at a given time (longer term once we have a cluster loader that should be the mechanism coordinating experiments in my opinion)
- history of what experiments we were doing in the repo :)

BTW - given that those "manual" suites weren't run even once, what I would suggest it switching them into:
- manual-gce
- manual-gke
[without -up and -down]
and starting using them for exactly that purpose

Thoughts?

Comment (shyamjvs) [reply test/e2e/scalability/load.go]: > However, in this particular case, your goal is to only discover the limits of iptables, and for that particular purpose the smaller cluster we can achieve it with, the better.

Sure, I understand. However, it's still unclear to me that the situation wrt iptables would remain similar in both scenarios. Just making up an e.g - having lesser #watches or smaller master may affect the rate at which watch events may get delivered which in turn may affect the iptables update operations. Another e.g could be that kube-proxy may be more resource-starved when there are 75 pods running per node instead of 30 (due to kubelet, etc).

> But I also don't think that changing our main (release-blocking) tests for that purpose is the best idea.

I agree, we shouldn't normally touch that one. I'm doing it the lazy way as it's a relatively "quiet" period now ;)

> manual test-suite

That's a nice idea. And in fact my thinking ~year ago - https://github.com/kubernetes/kubernetes/issues/50907
However, what I realized in time is that devs in general tend to find it hard to follow such a workflow (touching multiple places in code and then understanding prow details for manual trigger, etc). It's easier to achieve all the above using our large-scale presubmits, where triggering is just one single command.

**Comment by shyamjvs** [comment]
PTAL

/retest

File: test/e2e/scalability/load.go (hunk 1 above)
Comment (wojtek-t) [reply test/e2e/scalability/load.go]: Re 75-pods per node - yeah, that can have some impact, but since running test on 2k-node test is much much cheaper, let's try to:
- estimate the limit on 2k-node cluster (you will need to manually bump master size probably)
- then confirm on 5k-node one that it's the same (I suspect it to be roughly the same)


As discussed offline, presubmits sound even better.
Longer term, the caveats are that:
- ideally I would like to be able to specify when running command the size cluster, e.g. run-scale-tests 2000
- we need to work on converging setup for all different tests (i.e. change all tests to have exactly the same configuration no matter on size (e.g. logging, audit logging etc. is what differs)

File: test/e2e/scalability/load.go [hunk 2]
```diff
@@ -535,8 +535,21 @@ func GenerateConfigsForGroup(
...
 			ConfigMapNames: configMapNames,
-			// Define a label to group every 2 RCs into one service.
-			Labels: map[string]string{svcLabelKey: groupName + "-" + strconv.Itoa((i+1)/2)},
+		}
+
+		// We assume by default that every RC is part of its own service.
+		// However, you can override it using RC_TO_SERVICE_RATIO env var.
+		// For simplicity, we only allow integer values for the env var.
+		rcToServiceRatio := 1
+		if rcToServiceRatioEnv := os.Getenv("RC_TO_SERVICE_RATIO"); rcToServiceRatioEnv != "" {
```
Comment (wojtek-t) [inline test/e2e/scalability/load.go]: I just thought about it and I'm not convinced that this is the best way of doing that.
Let's say that I would like a service for 2/3rd of services - with this approach it's not possible.

I think we need something slightly more generic that 1/n-th.

Comment (shyamjvs) [reply test/e2e/scalability/load.go]: I know this sucks :)
Also exactly the reason why I added that comment `For simplicity, we only allow integer values for the env var`. But if you want I can try to enable floating point service:RC ratio. Does that sgty?

Comment (wojtek-t) [reply test/e2e/scalability/load.go]: My fear is that you will see that with 1/2 of backends everything will work and then you're stuck because with this code you can't really test e.g. 2/3rd backends.
So I think that floating point ration makes more sense here.

Comment (shyamjvs) [reply test/e2e/scalability/load.go]: Makes sense. Changed it, PTAL.

**Comment by shyamjvs** [comment]
I'm removing the milestone as it seems like code freeze is being lifted anyway by EOD today.
